x509/src/certificate.rs

```diff
@@ -0,0 +1,145 @@
+
+/// X.509 `TBSCertificate` as defined in [RFC 5280 Section 4.1.2.5]
+///
+/// ```text
+///   TBSCertificate  ::=  SEQUENCE  {
+///       version         [0]  Version DEFAULT v1,
+///       serialNumber         CertificateSerialNumber,
+///       signature            AlgorithmIdentifier{SIGNATURE-ALGORITHM, {SignatureAlgorithms}},
+///       issuer               Name,
+///       validity             Validity,
+///       subject              Name,
+///       subjectPublicKeyInfo SubjectPublicKeyInfo,
+///       ... ,
+///       [[2:               -- If present, version MUST be v2
+///       issuerUniqueID  [1]  IMPLICIT UniqueIdentifier OPTIONAL,
+///       subjectUniqueID [2]  IMPLICIT UniqueIdentifier OPTIONAL
+///       ]],
+///       [[3:               -- If present, version MUST be v3 --
+///       extensions      [3]  Extensions{{CertExtensions}} OPTIONAL
+///       ]], ... }
+/// ```
+/// [RFC 5280 Section 4.1.2.5]: https://datatracker.ietf.org/doc/html/rfc5280#section-4.1.2.5
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct TBSCertificate<'a> {
+    /// version         [0]  Version DEFAULT v1,
+    pub version: u8,
+    /// serialNumber         CertificateSerialNumber,
+    pub serial_number: UIntBytes<'a>,
+    /// signature            AlgorithmIdentifier{SIGNATURE-ALGORITHM, {SignatureAlgorithms}},
+    pub signature: AlgorithmIdentifier<'a>,
+    /// issuer               Name,
+    pub issuer: RDNSequence<'a>,
+    /// validity             Validity,
+    pub validity: Validity,
+    /// subject              Name,
+    pub subject: RDNSequence<'a>,
+    /// subjectPublicKeyInfo SubjectPublicKeyInfo,
+    pub subject_public_key_info: SubjectPublicKeyInfo<'a>,
+    /// issuerUniqueID  [1]  IMPLICIT UniqueIdentifier OPTIONAL,
+    pub issuer_unique_id: Option<BitString<'a>>,
+    /// subjectUniqueID [2]  IMPLICIT UniqueIdentifier OPTIONAL
+    pub subject_unique_id: Option<BitString<'a>>,
+    /// extensions      [3]  Extensions{{CertExtensions}} OPTIONAL
+    pub extensions: Option<Extensions<'a>>,
+}
+
+impl<'a> Decodable<'a> for TBSCertificate<'a> {
+    fn decode(decoder: &mut Decoder<'a>) -> der::Result<Self> {
+        decoder.sequence(|decoder| {
+            let version = decoder
+                .context_specific::<u8>(VERSION_TAG, TagMode::Explicit)?
+                .unwrap();
+            if version != VERSION {
+                return Err(der::Tag::Integer.value_error());
+            }
+
+            // TODO refactor unwrap if this implementation remains
+            let serial_number = UIntBytes::decode(decoder).unwrap();
+            let signature = AlgorithmIdentifier::decode(decoder).unwrap();
+            let issuer = RDNSequence::decode(decoder).unwrap();
+            let validity = Validity::decode(decoder).unwrap();
+            let subject = RDNSequence::decode(decoder).unwrap();
+            let subject_public_key_info = SubjectPublicKeyInfo::decode(decoder).unwrap();
```

Comment:
You should be able to replace pretty much all of these unwraps with `?`...

Reply:
Will need to refactor the version field unwrap too.